### DISCUSSION TITLE
Widgets

### DISCUSSION THREAD

**Discussion by dvarrazzo** [Textualize/textual#375] [Q&A]
Hello,

I have a command line tool which should be extended to allow users to choose among several settings, so I'm thinking to add an interactive mode, where settings would be chosen in menus via combo boxes, input boxes, etc. Very Windows 95 dialog forms. The users use Linux, Windows, macOS.

Would Textual be a good choice for this type of app? I haven't seen, in the example, any demo about widgets. Should I have to write my own check box class, in order to use Textual?

Thank you very much!

**Comment by USLTD** [comment]
Currently, **textual** doesn't have any input widgets and only third party library - [**textual-inputs**](https://github.com/sirfuzzalot/textual-inputs) is available which is experimental, so I recommend using other TUI library or write your own input widgets but remember that it may not work as intended since **textual**'s code and API is constantly changing.

**Reply by dvarrazzo** [reply]
Thank you very much. Is there any TUI you feel like suggesting for my use case? Something to create easily a simple, if boring, form, and to be multi-platform?

I will be happy to come back and check Textual once it is more stable :slightly_smiling_face:

**Reply by sanders41** [reply]
I have not used it myself, but I have see others do interesting TUIs with [curses](https://docs.python.org/3/howto/curses.html). I know it can accept text input, but I’m not sure about combo boxes, check boxes, etc.

**Reply by USLTD** [reply]
> Thank you very much. Is there any TUI you feel like suggesting for my use case? Something to create easily a simple, if boring, form, and to be multi-platform?
> 
> I will be happy to come back and check Textual once it is more stable 🙂

There are many variations and each of them have their own pros and cons.

Urwid
Blessings
Blessed
Notcurses (Python bindings)
Npyscreen

And more.

**Reply by dvarrazzo** [reply]
Nothing based on curses works on Windows, right?

- [Blessed](https://pypi.org/project/blessed/) seems to work on Windows, but it's about positions and colours, not widgets,
- [Blessing](https://github.com/erikrose/blessings) is older, Blessed forked from it, unmaintained for 2 years, no Windows
- [Npyscreen](https://pypi.org/project/npyscreen/#files) had no releases for 7 years. wrapper around curses
- [urwid](http://urwid.org/) doesn't work on windows
- notcurses is [not packaged for windows](https://pypi.org/project/notcurses/#files) and takes a [massive toolchain](https://github.com/dankamongmen/notcurses/blob/master/INSTALL.md#microsoft-windows) to build there - no `pip install`.

My question was, is there any TUI portable and widget-oriented around? I'd take this as a no.

**Reply by USLTD** [reply]
> Python-only libraries

There is also cross-platform [**nurses_2**](https://github.com/salt-die/nurses_2), successor to [**nurses**](https://github.com/salt-die/nurses), or [**Python Prompt Toolkit**](https://github.com/prompt-toolkit/python-prompt-toolkit).
There is also currently-Unix-only [**pyTermTK**](https://github.com/ceccopierangiolieugenio/pyTermTk), successor to [**pyCuT**](https://github.com/ceccopierangiolieugenio/pyCuT).

There are also other less-known TUI projects which you can search using **GitHub** or **google**.
